### PR TITLE
Add cache, merge linear layers, and etc.

### PR DESCRIPTION
**Pull request by ElectronicElephant**

### CODE REVIEW

File: nn_models.py [hunk 1]
```diff
@@ -117,14 +113,19 @@ def forward(self, x):
...
 
         return out1, out2, out3, out4, out5
 
 
 def resnet20():
     return ResNet(BasicBlock, [3, 3, 3])
+
+
+def resnet_min():
+    return ResNet(BasicBlock, [1, 1, 1])
```
Comment (PhotonQuantum) [inline nn_models.py:131]: 这个 resnet_min 可以正常工作吗？

Comment (ElectronicElephant) [reply nn_models.py:131]: It converged very slowly so I didn't fully test it.

Comment (PhotonQuantum) [reply nn_models.py:131]: alright, so we may safely delete this?

Comment (ElectronicElephant) [reply nn_models.py:131]: 没错!

**Comment by PhotonQuantum** [comment]
LGTM

**Pull request merged**